influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +125,213 @@ struct CacheValue {
+                    CacheEntryState::Fetching(tx) => tx.clone(),
+                    _ => bail!("attempted to set success state on a non-fetching cache entry"),
+                };
+                let cache_value = Arc::new(value);
+                entry.state = CacheEntryState::Success(Arc::clone(&cache_value));
+                entry
+                    .hit_time
+                    .store(self.time_provider.now().timestamp_nanos(), Ordering::SeqCst);
+                // TODO(trevor): what if size is greater than cache capacity?
+                let additional = entry.size();
+                self.used.fetch_add(additional, Ordering::SeqCst);
+                // notify any watching requests that tried to get this entry while it was
+                // fetching:
+                let _ = tx.send(Some(cache_value));
+                Ok(())
+            }
+            Entry::Vacant(_) => bail!("attempted to set success state on an empty cache entry"),
+        }
+    }
+
+    /// Remove an entry from the cache, as well as its associated size from the used capacity
+    fn remove(&self, path: &Path) {
+        let Some((_, entry)) = self.map.remove(path) else {
+            return;
+        };
+        self.used.fetch_sub(entry.state.size(), Ordering::SeqCst);
+    }
+
+    /// A cache should be pruned if it is using all of its memory capacity
+    fn should_prune(&self) -> bool {
+        let used = self.used.load(Ordering::SeqCst);
+        used >= self.capacity
+    }
+
+    /// Prune least recently hit entries from the cache
+    fn prune(&self) {
+        let n_to_prune = (self.map.len() as f64 * self.prune_percent).floor() as usize;
+        // use a BinaryHeap to determine the cut-off time, at which, entries that were
+        // last hit before that time will be pruned:
+        let mut time_heap = BinaryHeap::with_capacity(n_to_prune);
+
+        for map_ref in self.map.iter() {
+            let hit_time = map_ref.value().hit_time.load(Ordering::Relaxed);
+            if time_heap.len() < n_to_prune {
+                // if the heap isn't full yet, throw this time on:
+                time_heap.push(hit_time);
+            } else if hit_time < *time_heap.peek().unwrap() {
+                // otherwise, the heap is at its capacity, so only push if the hit_time
+                // in question is older than the top of the heap (after pop'ing the top
+                // of the heap to make room)
+                time_heap.pop();
+                time_heap.push(hit_time);
+            }
+        }
 
-impl WeightScale<Path, CacheEntry> for CacheEntryScale {
-    fn weight(&self, key: &Path, value: &CacheEntry) -> usize {
-        key.as_ref().len() + value.size()
+        // track the total size of entries that get freed:
+        let mut freed = 0;
+        // drop entries with hit times before the cut-off:
+        let cutoff_time = *time_heap.peek().unwrap();
+        self.map.retain(|_, entry| {
```

Comment:
I will try this again since I am using DashMap. Previously, I ran into a borrow checker issue where the borrowed path `&'a Path` in `Deletable<'a>` prevents the mutable reference needed to call `remove`, which takes `&mut self`. DashMap doesn't require mutability, but I feel like it might just deadlock instead. We could also just clone the path into `Deletable` and tradeoff space for time.

Reply:
> Ah - I see, can Deletables hold Arc<str> maybe? - under the hood Path looks like a wrapper around String anyway and implements AsRef<str>. So path.as_ref() could do?

I think this works reasonably well, see https://github.com/influxdata/influxdb/pull/25389/commits/ba5ff2b19127d9290c885d943651793fd185d102